sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -379,364 +401,248 @@ class GpuDynamicPartitionDataSingleWriter(
+   * Returns a tuple with an array of the splits as `ContiguousTable`'s, an array of
+   * paths and bucket ids to use to write each partition and(or) bucket file.
    */
-  private def splitBatchByKeyAndClose(
-      batch: ColumnarBatch,
-      partDataTypes: Array[DataType]): Array[SplitAndPath] = {
-    val (outputColumnsBatch, partitionColumnsBatch) = withResource(batch) { _ =>
-      closeOnExcept(getOutputColumnsAsBatch(batch)) { outputColumnsBatch =>
-        closeOnExcept(getPartitionColumnsAsBatch(batch)) { partitionColumnsBatch =>
-          (outputColumnsBatch, partitionColumnsBatch)
+  private def splitBatchByKeyAndClose(batch: ColumnarBatch): Array[SplitPack] = {
+    val cbs@Array(keysCb, dataCb) = withResource(batch) { _ =>
+      Array(getKeysBatch(batch), getDataColumnsAsBatch(batch))
+    }
+    val (keyHostCb, splits) = withResource(cbs) { _ =>
+      val keysCbTypes = GpuColumnVector.extractTypes(keysCb)
+      val (distinctKeysTbl, splitIds) = withResource(keysCb) { _ =>
+        withResource(GpuColumnVector.from(keysCb)) { keysTable =>
+          val distinctKeysTbl = distinctAndSort(keysTable)
+          (distinctKeysTbl, splitIndexes(keysTable, distinctKeysTbl))
         }
       }
-    }
-    val (cbKeys, partitionIndexes) = closeOnExcept(outputColumnsBatch) { _ =>
-      val partitionColumnsTbl = withResource(partitionColumnsBatch) { _ =>
-        GpuColumnVector.from(partitionColumnsBatch)
+      cbs(0) = null // avoid double close
+      val keyHostCb = withResource(distinctKeysTbl) { _ =>
+        copyToHostAsBatch(distinctKeysTbl, keysCbTypes)
       }
-      withResource(partitionColumnsTbl) { _ =>
-        withResource(distinctAndSort(partitionColumnsTbl)) { distinctKeysTbl =>
-          val partitionIndexes = splitIndexes(partitionColumnsTbl, distinctKeysTbl)
-          val cbKeys = copyToHostAsBatch(distinctKeysTbl, partDataTypes)
-          (cbKeys, partitionIndexes)
-        }
-      }
-    }
-
-    val splits = closeOnExcept(cbKeys) { _ =>
-      val spillableOutputColumnsBatch =
-        SpillableColumnarBatch(outputColumnsBatch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-      withRetryNoSplit(spillableOutputColumnsBatch) { spillable =>
-        withResource(spillable.getColumnarBatch()) { outCb =>
-          withResource(GpuColumnVector.from(outCb)) { outputColumnsTbl =>
-            withResource(outputColumnsTbl) { _ =>
-              outputColumnsTbl.contiguousSplit(partitionIndexes: _*)
+      closeOnExcept(keyHostCb) { _ =>
+        val scbOutput = SpillableColumnarBatch(dataCb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+        val splits = withRetryNoSplit(scbOutput) { scb =>
+          withResource(scb.getColumnarBatch()) { outCb =>
+            withResource(GpuColumnVector.from(outCb)) { outputColumnsTbl =>
+              withResource(outputColumnsTbl) { _ =>
+                outputColumnsTbl.contiguousSplit(splitIds: _*)
+              }
             }
           }
         }
+        cbs(1) = null
```

Comment:
Again who closed this column? it is not clear form the code what is happening here.

Reply:
Did some refactor and removed this.